nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -474,7 +474,9 @@ public static DataType inferDataType(final Object value, final DataType defaultT
 
         // A value of a Map could be either a Record or a Map type. In either case, it must have Strings as keys.
         if (value instanceof Map) {
-            final Map<String, ?> map = (Map<String, ?>) value;
+            Map<?, ?> m = (Map<?, ?>) value;
```

Comment:
In the majority of cases, the Map will already have Strings for keys. Might recommend first checking this, in order to avoid the garbage collection & related resource consumption when it's usually not needed. Something along the lines of:
```
final Map<String, Object> map;
boolean allStrings = true;
for (final Object key : ((Map<?, ?) value).keySet()) {
  if (!(key instanceof String)) {
    allStrings = false;
    break;
  }
}

if (allKeyStrings) {
  map = (Map<String, Object>) value;
} else {
  final Map<?, ?> m = (Map<?, ?>) value;
  map = new HashMap<>(m.size());
  m.forEach((k, v) -> map.put(k == null ? null : k.toString(), v));
}
```

Is also worth noting that Map does allow for null keys. I don't believe in our uses that we'd ever encounter this to this point, but it does make sense to check for it and avoid an NPE if it does occur.